src/jvm/main/org/jetbrains/kotlinx/lincheck/verifier/CachedVerifier.java

```diff
@@ -25,9 +25,13 @@ public abstract class CachedVerifier implements Verifier {
 
     @Override
     public boolean verifyResults(ExecutionScenario scenario, ExecutionResult results) {
-        boolean newResult = previousResults.computeIfAbsent(scenario, s -> new HashSet<>()).add(results);
-        if (!newResult) return true;
-        return verifyResultsImpl(scenario, results);
+        Set<ExecutionResult> executionResults = previousResults.computeIfAbsent(scenario, s -> new HashSet<>());
+        if (executionResults.contains(results)) return true;
+        boolean isValid = verifyResultsImpl(scenario, results);
+        if (isValid) {
+            executionResults.add(results);
```

Comment:
Why do you need this change?

Reply:
Because `executionResults ` should contain only valid results. While before it cached all the results and incorrectly returned `true` when called repeadetly

Reply:
Why is it becoming an issue now?

Reply:
Because previously `CachedVerifier` was never used after incorrect result is found

Reply:
What is the problem now?

Reply:
Now, we re-use the same verifier after an error is found, and with the old logic, it would have found the same wrong result in the cache and indicated that execution is correct, while it's not. So, I absolutely agree that this change is necessary. I reflected it in the comments.

Reply:
Thank you, now it is clear.